File: euler151.py
def cut(a2, a3, a4, a5):
    sheets = float(a2+a3+a4+a5)
    if sheets==0: return 0
    singles = 1 if sheets==1 and a5==0 else 0
    if a2: singles += a2*cut(a2-1, a3+1, a4+1, a5+1)
    if a3: singles += a3*cut(a2  , a3-1, a4+1, a5+1)
    if a4: singles += a4*cut(a2  , a3  , a4-1, a5+1)
    if a5: singles += a5*cut(a2  , a3  , a4  , a5-1)
    return singles/sheets
print round(cut(1, 1, 1, 1), 6)



Example,
(a2,a3,a4,a5)
in case of (0,1,0,1)
if (a3) and if(a5) will be true, and called.

so s will consist of the returns of the calls :
singles = 1 if sheets==1 and a5==0 else 0
a3 * (a2,a3-1,a4+1,a5+1)  = 1 * (0,0,1,2)
a5 * (a2,a3,a4,a5-1)      = 1 * (0,1,0,0)



Our function goes through every possible occurence.
It does this recursively,
It returns the chance of us reaching a single sheet
From the last state. This would normally also count the
occurences of a single a5 paper, but we exclude that In
our If -statement at the  3th line of the function.

at every call it calculates the chance of ending up at
a single-sheet occurence in the rest of the tree.
It does this by
in the end the whole function, returns the chance of us
reaching a single sheet, in the whole tree of possibilities.

File: euler151.hs
import Text.Printf

main :: IO ()
main = do
  printf "%.6f\n" (cut(1,1,1,1))
  printf "%.6f\n" (fun(1,1,1,1))

fun ::(Double,Double,Double,Double) -> Double
fun (0,0,0,1) = 0
fun (0,0,1,0) = fun (0,0,0,1) + 1
fun (0,1,0,0) = fun (0,0,1,1) + 1
fun (1,0,0,0) = fun (0,1,1,1) + 1
fun (a,b,c,d) =
    (a2 + a3 + a4 + a5) / (a + b + c + d)
    where
    a2 | a > 0 = a * fun (a-1,b+1,c+1,d+1)
          | otherwise = 0
    a3 | b > 0 = b * fun (a,b-1,c+1,d+1)
          | otherwise = 0
    a4 | c > 0 = c * fun (a,b,c-1,d+1)
          | otherwise = 0
    a5 | d > 0 = d * fun (a,b,c,d-1)
          | otherwise = 0

cut :: (Float,Float,Float,Float) -> Float
cut(0,0,0,1)     = 0
cut(a2,a3,a4,a5) = (s2+s3+s4+s5 + single) / sum [a2,a3,a4,a5]
  where
    single | sum [a2,a3,a4,a5] == 1 = 1
           | otherwise = 0
    s2     | a2 > 0 = a2 * cut(a2-1,a3+1,a4+1,a5+1)
           | otherwise = 0
    s3     | a3 > 0 = a3 * cut(a2  ,a3-1,a4+1,a5+1)
           | otherwise = 0
    s4     | a4 > 0 = a4 * cut(a2  ,a3  ,a4-1,a5+1)
           | otherwise = 0
    s5     | a5 > 0 = a5 * cut(a2  ,a3  ,a4  ,a5-1)
           | otherwise = 0
